File: n_prime_numbers/n_prime_numbers.py
import numpy as np
import time

def is_prime_linear(num):
    if num <= 1:
        return False

    for i in range(2, (num//2)+1):
        if num % i == 0:
            return False
    return True

def is_prime_sqrt(num):
    if num <= 1:
        return False

    for i in range(2,int(np.sqrt(num))+1):
        if num % i == 0:
            return False
    return True

def find_primes(n):
    primes = []
    for i in range(2,n+1):
        if is_prime_sqrt(i):
            primes.append(i)

    return primes

def find_primes_vec(n):
    primes = []
    is_prime = True
    for i in range(2,n+1):
        is_prime = True
        i_sq = np.sqrt(i)
        for prime in primes:
            if i % prime == 0:
                is_prime = False
                break
            if prime > i_sq:
                break
        if is_prime:
            primes.append(i)
    return primes

if __name__ == "__main__":
    # n = 1000000
    n = 1000000
    start_time = time.time()
    # prime_list = find_primes(n)
    prime_list = find_primes_vec(n)
    end_time = time.time()
    
    # print("The prime numbers upto N are:", prime_list)
    print("Time to find primes: ", round(end_time - start_time, 6), "seconds")

File: n_prime_numbers/n_prime_numbers.cpp
#include <bits/types/clock_t.h>
#include<iostream>
#include <numbers>
#include<vector>
#include<ostream>
#include<ctime>
#include<math.h>

using namespace std;
bool is_prime_linear(int num)
{
  if(num <= 1)
    return false;

  for(int i = 2; i <= num/2; i++)
    if(num%i == 0)
      return false;

  return true;
}

bool is_prime_sqrt(int num)
{
  if(num <= 1)
    return false;

  for(int i = 2; i <= int(std::sqrt(num)); i++)
    if(num%i == 0)
      return false;

  return true;
}

std::vector<int> find_primes(int n)
{
    std::vector<int> primes_list;
    for(int i = 2; i <= n; i++)
    {
      if(is_prime_sqrt(i))
        primes_list.push_back(i);
    }
    return primes_list;
}

std::vector<int> find_primes_vec(int n)
{
  std::vector<int> primes_list;
  bool is_prime = true;
  int i_sqrt;
  for(int i = 2; i <= n; i++)
  {
    is_prime = true;
    i_sqrt = std::sqrt(i);
    for(auto it = primes_list.begin(); it != primes_list.end(); it++)
    {
      if(i % (*it) == 0 || (*it) > i_sqrt)
      {
        is_prime = false;
        break;
      }
    }
    if(is_prime)
      primes_list.push_back(i);
  }
  return primes_list;
}

int main()
{
  std::clock_t start_time = std::clock();
  // std::vector<int> primes = find_primes(10000000);
  std::vector<int> primes = find_primes_vec(10000000);
  std::clock_t end_time = std::clock();

  std::cout<<"The list of primes is: " << "\n";
  // for(auto it = primes.begin(); it != primes.end(); it++)
  //   std::cout<< *it << " ";

  double elapsed_time = double(end_time - start_time) / CLOCKS_PER_SEC;

  std::cout<< "\nTime taken by the program: " << elapsed_time << "\n";

  return 0;
}